pkg/network/iface/link.go

```diff
@@ -0,0 +1,227 @@
+package iface
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/vishvananda/netlink"
+	"k8s.io/klog"
+	"k8s.io/utils/exec"
+	"k8s.io/utils/net/ebtables"
+)
+
+const defaultPVID = uint16(1)
+
+type Link struct {
+	link   netlink.Link
+	addr   []netlink.Addr
+	routes []netlink.Route
+}
+
+func GetLink(name string) (*Link, error) {
+	if name == "" {
+		return nil, fmt.Errorf("link name could not be empty string")
+	}
+
+	l, err := netlink.LinkByName(name)
+	if err != nil {
+		return nil, fmt.Errorf("could not lookup link, error: %w, link: %s", err, name)
+	}
+	addr, err := netlink.AddrList(l, netlink.FAMILY_V4)
+	if err != nil {
+		return nil, fmt.Errorf("list IPv4 address of %s failed, error: %w", l.Attrs().Name, err)
+	}
+	routes, err := netlink.RouteList(l, netlink.FAMILY_V4)
+	if err != nil {
+		return nil, fmt.Errorf("list routes of %s failed, error: %w", l.Attrs().Name, err)
+	}
+
+	return &Link{
+		link:   l,
+		addr:   addr,
+		routes: routes,
+	}, nil
+}
+
+// AddBridgeVlan adds a new vlan filter entry
+// Equivalent to: `bridge vlan add dev DEV vid VID master`
+func (l *Link) AddBridgeVlan(vid uint16) error {
+	if vid == defaultPVID {
+		return nil
+	}
+
+	if err := netlink.BridgeVlanAdd(l.link, vid, false, false, false, true); err != nil {
+		return fmt.Errorf("add iface vlan failed, error: %v, link: %s, vid: %d", err, l.Name(), vid)
+	}
+
+	return nil
+}
+
+// DelBridgeVlan adds a new vlan filter entry
+// Equivalent to: `bridge vlan del dev DEV vid VID master`
+func (l *Link) DelBridgeVlan(vid uint16) error {
+	if vid == defaultPVID {
+		return nil
+	}
+
+	if err := netlink.BridgeVlanDel(l.link, vid, false, false, false, true); err != nil {
+		return fmt.Errorf("delete iface vlan failed, error: %v, link: %s, vid: %d", err, l.Name(), vid)
+	}
+
+	return nil
+}
+
+func (l *Link) SetMaster(br *Bridge) error {
+	if l.link.Attrs().MasterIndex == br.bridge.Index {
+		return nil
+	}
+
+	klog.Infof("%s set master as %s", l.Name(), br.Name())
+	return netlink.LinkSetMaster(l.link, br.bridge)
+}
+
+func (l *Link) SetNoMaster() error {
+	if l.LinkAttrs().MasterIndex == 0 {
+		return nil
+	}
+
+	down := make(chan bool)
+	var isUp bool
+	if l.LinkAttrs().OperState == netlink.OperUp {
+		isUp = true
+		go l.reportDown(down)
+	}
+
+	klog.Infof("%s set no master", l.Name())
+	if err := netlink.LinkSetNoMaster(l.link); err != nil {
+		return err
+	}
+
+	if isUp {
+		select {
+		case <-down:
+			return netlink.LinkSetUp(l.link)
+		case <-time.After(time.Minute):
+			klog.Infof("Waiting for link down event timeout")
+		}
+	}
+
+	return nil
+}
+
+func (l *Link) reportDown(down chan bool) {
+	linkCh := make(chan netlink.LinkUpdate)
+	done := make(chan struct{})
+	if err := netlink.LinkSubscribe(linkCh, done); err != nil {
```

Comment:
Can we do `netlink.LinkSubscribe(linkCh, down)` so that we don't need additional done channel